tests/unit/test_util.py

```diff
@@ -235,6 +259,22 @@ def test_sample_batch_indices(
                 num_batches=num_batches,
             )
 
+    def test_sample_batch_indices_uniqueness(self) -> None:
+        """
+        Test that sample_batch_indices produces row-unique indices.
+        """
+        data_size = 100
+        batch_size = 50
+        num_batches = 10
+        batch_indices = sample_batch_indices(
+            random_key=key(0),
+            data_size=data_size,
+            batch_size=batch_size,
+            num_batches=num_batches,
+        )
+        for row in batch_indices:
```

Comment:
We currently have an intra-row check (uniqueness of indices within the row), but not an inter-row uniqueness check (uniqueness between the rows). I assume this is because inter-row uniqueness is not guaranteed. I.E we would accept `[1,2],[1,2],[2,1],[1,2]` for four batches of size two, where data_size is also two. My only concern is that the test currently permits a function that generates the same order of indices for every batch.

I think the probability of sampling any specific (ordered) batch of indices is `1 / math.perm(data_size, batch_size)`; with the parameters in this test, the probability is on the order of 1e-94 (the parameters could be chosen to make this probability arbitrarily small). I believe this allows us to reasonably expect all rows to be unique (perhaps leave a comment in the test explaining why this is expected).

The test for the above could be something like `assert jnp.unique(batch_indices, axis=0).shape[0] == num_batches` . This will implicitly check the shape of `batch_indices` if you don't want a dedicate shape assertion.

Reply:
Added tests that cover these cases